pkgs/native_assets_cli/lib/src/locking/locking.dart

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:logging/logging.dart';
+
+/// Run [callback] with this Dart process having exclusive access to
+/// [directory].
+///
+/// Note multiple isolates and isolate groups in the same Dart process share
+/// locks, so these will be able to enter the exclusive section simultanously.
+///
+/// If provided, the [timeout] parameter determines how long to retry before
+/// giving up. If the [timeout] is exceeded a [TimeoutException] is thrown.
+///
+/// If provided, the [logger] streams information on the the locking status, and
+/// also streams error messages.
+Future<T> runUnderDirectoryLock<T>(
+  Directory directory,
+  Future<T> Function() callback, {
+  Duration? timeout,
+  Logger? logger,
+}) async {
+  const lockFileName = '.lock';
+  final lockFile = File.fromUri(directory.uri.resolve(lockFileName));
+  return _runUnderFileLock(
+    lockFile,
+    callback,
+    timeout: timeout,
+    logger: logger,
+  );
+}
+
+/// Run [callback] with this Dart process having exclusive access to [file].
+///
+/// Note multiple isolates and isolate groups in the same Dart process share
+/// locks, so these will be able to enter the exclusive section simultanously.
+///
+/// If provided, the [timeout] parameter determines how long to retry before
+/// giving up. If the [timeout] is exceeded a [TimeoutException] is thrown.
+///
+/// If provided, the [logger] streams information on the the locking status,
+/// and also streams error messages.
+Future<T> _runUnderFileLock<T>(
+  File file,
+  Future<T> Function() callback, {
+  Duration? timeout,
+  Logger? logger,
+}) async {
+  if (!await file.exists()) await file.create(recursive: true);
+  final randomAccessFile = await file.open(mode: FileMode.write);
```

Comment:
If Windows does not allow another file descriptor to write to a locked file, and `FileMode.write` says "The file is overwritten if it already exists. The file is created if it does not already exist.", shouldn't this *fail* if the file is already locked? At least on Windows?

(If not, why not?)

Reply:
This case is covered by multiple tests. They explicitly lock the file, and then spawn a new process that executes exactly this code. I believe opening the file in write mode (which gives you a file handle) is fine. It's the write operation with such file handle that will fail. (You need a file handle to do the locking call to begin with, so it would be weird if getting the file handle would fail.)